### PR TITLE
Proposal for scaling down from fully connected affinity

### PR DESCRIPTION
**Pull request by kriskowal**
cc @mranney @jcorbin @Raynos

### CODE REVIEW

File: docs/affinity.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
+
+# Affinity
+
+For every service, there is a set of service workers and a set of affine
```
Comment (jcorbin) [inline docs/affinity.md]: suggest "and a set of fine (sic) relays" ;-)

File: docs/affinity.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+Hyperbahn relays (the subset of the relay ring that is responsible for traffic
+to and from the given service).
+The minimum viable Hyperbahn fully connects the affine workers and relays,
```
Comment (jcorbin) [inline docs/affinity.md]: I think you have "affine" modifying the wrong noun here, shouldn't it be: "connects the workers to the affine relays" ?

File: docs/affinity.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+however over time, the relay workers encroach on the maximum file descriptor
```
Comment (jcorbin) [inline docs/affinity.md]: don't say "relay workers", simply say "relays" for clarity... "workers" evokes the service workers

File: docs/affinity.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+limit.
+The following is a strategy for determining which service workers each relay
+should connect to.
+
+There is a value *n* that is the number of connections each service worker
+should retain to the Hyperbahn.
+This number should never be less than 3 to ensure some measure of redundancy.
+THere is also a value *m* that is the minimum number of connections each relay
```
Comment (jcorbin) [inline docs/affinity.md]: `s/THere/there/`

File: docs/affinity.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+resources (possibly including file descriptors, CPU, and throughput) consumed
+by each relay.
+
+By virtue of gossiping advertisements for each service to every member of the
+affinity set, every Hyperbahn relay knows the host and port of every worker for
+its affine services.
+*W* is the number of workers for the service, and is known by every affine
+relay.
+
+Each relay within the afinity set for a service can know its position, *i*, by
```
Comment (jcorbin) [inline docs/affinity.md]: missing an f: `s/afinity/affinity/`

**Comment by kriskowal** [comment]
Feedback addressed, and I’ve expanded all of the single letter variable names.

**Comment by Raynos** [comment]
I like this approach; :+1:

**Comment by kriskowal** [comment]
@raynos, This is the result of a simulated partition of 100 relays to 100 workers. A small subset of relays get to retain some or all of their peers. Most lose all 6 peers and gain 3 new ones. This is clearly not ideal. Something something hashes something something maybe.

```
--- w=100 r=100 ---
--- w=100 r=50 ---
0: 6 +[3,4,5] -[]
1: 6 +[4,5,6,7] -[1]
2: 6 +[8,9,5,6,7] -[2,3]
3: 6 +[6,7,8,9,10,11] -[3,4,5]
4: 6 +[8,9,10,11,12,13] -[4,5,6]
5: 6 +[10,11,12,13,14,15] -[5,6,7]
6: 6 +[12,13,14,15,16,17] -[8,6,7]
7: 6 +[14,15,16,17,18,19] -[8,9,7]
8: 6 +[16,17,18,19,20,21] -[8,9,10]
9: 6 +[18,19,20,21,22,23] -[9,10,11]
10: 6 +[20,21,22,23,24,25] -[10,11,12]
11: 6 +[22,23,24,25,26,27] -[11,12,13]
12: 6 +[24,25,26,27,28,29] -[12,13,14]
13: 6 +[26,27,28,29,30,31] -[13,14,15]
14: 6 +[32,33,28,29,30,31] -[16,14,15]
15: 6 +[32,33,34,35,30,31] -[16,17,15]
16: 6 +[32,33,34,35,36,37] -[16,17,18]
17: 6 +[34,35,36,37,38,39] -[17,18,19]
18: 6 +[36,37,38,39,40,41] -[18,19,20]
19: 6 +[38,39,40,41,42,43] -[19,20,21]
20: 6 +[40,41,42,43,44,45] -[20,21,22]
21: 6 +[42,43,44,45,46,47] -[21,22,23]
22: 6 +[44,45,46,47,48,49] -[24,22,23]
23: 6 +[46,47,48,49,50,51] -[24,25,23]
24: 6 +[48,49,50,51,52,53] -[24,25,26]
25: 6 +[50,51,52,53,54,55] -[25,26,27]
26: 6 +[52,53,54,55,56,57] -[26,27,28]
27: 6 +[54,55,56,57,58,59] -[27,28,29]
28: 6 +[56,57,58,59,60,61] -[28,29,30]
29: 6 +[58,59,60,61,62,63] -[29,30,31]
30: 6 +[64,65,60,61,62,63] -[32,30,31]
31: 6 +[64,65,66,67,62,63] -[32,33,31]
32: 6 +[64,65,66,67,68,69] -[32,33,34]
33: 6 +[66,67,68,69,70,71] -[33,34,35]
34: 6 +[68,69,70,71,72,73] -[34,35,36]
35: 6 +[70,71,72,73,74,75] -[35,36,37]
36: 6 +[72,73,74,75,76,77] -[36,37,38]
37: 6 +[74,75,76,77,78,79] -[37,38,39]
38: 6 +[76,77,78,79,80,81] -[40,38,39]
39: 6 +[78,79,80,81,82,83] -[40,41,39]
40: 6 +[80,81,82,83,84,85] -[40,41,42]
41: 6 +[82,83,84,85,86,87] -[41,42,43]
42: 6 +[84,85,86,87,88,89] -[42,43,44]
43: 6 +[86,87,88,89,90,91] -[43,44,45]
44: 6 +[88,89,90,91,92,93] -[44,45,46]
45: 6 +[90,91,92,93,94,95] -[45,46,47]
46: 6 +[96,97,92,93,94,95] -[48,46,47]
47: 6 +[96,97,98,99,94,95] -[48,49,47]
48: 6 +[96,97,98,99,0,1] -[48,49,50]
49: 6 +[0,1,98,99,2,3] -[49,50,51]
```

**Comment by Raynos** [comment]
Ping. Is this still reality.

If so lets merge it.

**Pull request merged**